### PR TITLE
✨ Allow webhook changes to OpenStackCluster.Spec.Bastion

### LINKED ISSUE #1054: Bastion host is immutable on 0.5.0 while it wasnt on 0.4.0
/kind bug

On a `OpenStackCluster` I ussualy config the bastion specs and set `enable` to `false`. I toggle the enable flag during troubleshooting or checking various components on the nodes. And i disable it again. This is rather useful to not waste capacity when its not required and also not have a backdoor to the nodes when its not required either. 

**What steps did you take and what happened:**
- upgrade to CAPO `0.5.0` and rest of the components to `1.0.1`
- toggle the bastion enable flag to `true`
- error as it is immutable

**What did you expect to happen:**
- create the bastion host as this was possible on 0.4.0

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): `0.5.0`
- Cluster-API version: `1.0.1`
- OpenStack version: `Stein`
- Kubernetes version (use `kubectl version`): many, shouldnt affect it
- OS (e.g. from `/etc/os-release`): ubuntu20.04, shouldnt affect it either

### PR DESCRIPTION
**Pull request by tobiasgiese**
**What this PR does / why we need it**:

This PR allows `OpenStackCluster.Spec.Bastion` to be updated (i.e., to be mutable).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1054

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

<sub>Tobias Giese <tobias.giese@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

### CODE REVIEW

File: api/v1alpha4/openstackcluster_webhook.go
Comment (seanschneeweiss) [inline api/v1alpha4/openstackcluster_webhook.go:26]: Not sure, but should v1alpha4 of CAPO use CAPI v1alpha4?
```suggestion
	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
```

Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook.go:26]: For CAPO v0.5.0 we‘re using CAPI v1beta1 (see https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1024#issuecomment-968406388)

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:71]: These are the same. Did you intend to change IdentityRef.Name?

File: api/v1beta1/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1beta1/openstackcluster_webhook_test.go:71]: As for v1alpha4

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:101]: This is an interesting one. I note that in this unit test we're changing both the bastion spec and the enabled flag. This is fine: we will create a new Bastion with the new spec.

However, note we also permit changing the bastion spec without flipping enabled, but this won't work. It won't fail, but the bastion will not be updated until the enabled flag is flipped to false and the bastion is deleted.

What do we want to do about this? Options:

1. Ignore it
2. Document it
3. Don't allow changing bastion spec if status.Bastion != nil
4. Make it work

For now I feel like 2 + 3 would be user friendly and relatively simple.

**Review by mdbooth — CHANGES_REQUESTED** [review]
Thanks for the unit tests! Couple of discussion points.

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:71]: You're totally right. This is a leftover from a previous test, but I will add the `IdentityRef.Name` 👍🏻

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:101]: I totally agree that we should document this behavior.
Regarding 3: do you think it's worth the effort to update our `reconcileBastion` to delete the bastion if sth has changed? Or is it enough to only allow changes if the enabled flag is false? Both are fine for me, I just think from the customer's perspective.
WDYT?

Comment (mdbooth) [reply api/v1alpha4/openstackcluster_webhook_test.go:101]: I feel like automatically deleting and recreating the bastion if something changed would be the most user friendly, but I should strongly caveat that with: I don't feel like I understand all the potential use cases here. Have we firmly set an expectation that we can blow the bastion away? Nobody's running anything on the bastion which is going to suffer data loss? I'd hope not.

I also think it would be fine to do 3 and do 4 later. I don't think they're incompatible: if the user writes orchestration to delete and recreate the bastion if they need to change it, that will continue to work even if we remove the requirement to do it. We're also being nice to them in the meantime by pointing out that it doesn't work.

Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:101]: Okay, I'll change my PR to 2 + 3 👍🏻
Thanks!

Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:123]: I updated my PR, can you PTAL @mdbooth? 🙂 
Documentation will be pushed during the day.

File: api/v1alpha4/openstackcluster_webhook.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook.go:84]: I *think* this is ok. Changing IdentityRef.Name is definitely a potential foot gun, but:
* I can't think of any reason to strictly forbid it
* There may be edge cases when it's useful to support allowing changing it

I'd appreciate a second opinion from anybody familiar with the original design decisions!

Comment (chrischdi) [reply api/v1alpha4/openstackcluster_webhook.go:84]: I was not part of the original design decisions but I also think it should be ok to allow changes.

E.g. maybe somebody wants to keep the old credential information and creating a new secret for the new credential.

Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook.go:84]: CAPA is only disallowing the removal of a IdentityRef

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/api/v1beta1/awscluster_webhook.go#L136-L142

Comment (sbueringer) [reply api/v1alpha4/openstackcluster_webhook.go:84]: IIRC the identityref is only used during reconcile to get the OpenStack creds and then access the OpenStack API with them. (aka it is not used to store them somewhere in the cluster).

If I'm correct, I see no reason to deny changes to the identityRef

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:141]: Added the documentation as well. Thx

File: api/v1alpha4/openstackcluster_webhook.go
Comment (mdbooth) [reply api/v1alpha4/openstackcluster_webhook.go]: That's also my understanding.

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:35]: nit: The tests names confused me, and perhaps are still confusing me a little. To me, mutability/immutability is a fundamental property of a thing. 'Immutable' to me means: "no changes at all, ever", whereas this is more like: "this specific change is not allowed".

For this specific test I might have written "Changing OpenStackCluster.Spec.IdentityRef.Kind is not allowed".

This is a nit! Please don't respin for this.

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:188]: I like this name better 😀

File: api/v1alpha4/openstackcluster_webhook.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook.go]: Won't this prevent setting Enabled=false once the bastion is deployed? i.e. we can never undeploy it in order to make changes?

**Review by mdbooth — CHANGES_REQUESTED** [review]
Side quest looks complete, but I think you've got a new problem in the main story line 😂

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go]: Updated the naming. WDYT?

File: api/v1alpha4/openstackcluster_webhook.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook.go]: Good catch 🙂 fixed it

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:125]: fixed

File: api/v1alpha4/openstackcluster_webhook.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook.go]: I must have lost sight of the flag somehow 😅

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:35]: Thank you! 🙏

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:231]: I think you're missing an interesting test here: flipping Enabled when the Bastion exists in status is allowed.

**Review by mdbooth — APPROVED** [review]
The clean unit tests make this so much easier to review, thanks!

I think there's an extra unit test you could add, but either way this
/lgtm

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (tobiasgiese) [reply api/v1alpha4/openstackcluster_webhook_test.go:231]: Good point. I added also another test `Disabling and changing the OpenStackCluster.Spec.Bastion while it's running is allowed`. This should be fine, as the bastion host will be destroyed anyway.

File: api/v1alpha4/openstackcluster_webhook_test.go
Comment (mdbooth) [inline api/v1alpha4/openstackcluster_webhook_test.go:233]: 👍

**Review by mdbooth — APPROVED** [review]
/lgtm

Thanks for the extra test. As discussed in Slack, I was worried that allowing changes to the spec before the bastion had actually been removed could potentially introduce a race which somebody might hit if, e.g. they scripted it. This approach feels simple and correct.

**Comment by mdbooth** [comment]
@tobiasgiese Tests are all passing. Do you want to remove the hold?

**Comment by tobiasgiese** [comment]
> @tobiasgiese Tests are all passing. Do you want to remove the hold?

I‘m still waiting for the approve of a CAPO maintainer. Prow won‘t merge until it‘s approved.

**Comment by tobiasgiese** [comment]
ptal @hidekazuna @jichenjc 🙂 thanks

**Comment by mdbooth** [comment]
> > @tobiasgiese Tests are all passing. Do you want to remove the hold?
> 
> I‘m still waiting for the approve of a CAPO maintainer. Prow won‘t merge until it‘s approved.

That's ok. If you remove the /hold it will merge immediately after it's approved.

**Comment by tobiasgiese** [comment]
> That's ok. If you remove the /hold it will merge immediately after it's approved.

I know that 🙂 we're using Prow at work.
I like to know when my PR is going to be merged. Also, I don't like to handover the responsibility (also as it's sometimes a approve-blocker for someone).

**Comment by seanschneeweiss** [comment]
/LGTM
Agree very much that the tests make it easy to review 👍

**Comment by hidekazuna** [comment]
@tobiasgiese You can remove /hold if you want 😃

**Comment by tobiasgiese** [comment]
thx @hidekazuna :)

/unhold

**Pull request merged**
